base/namedtuple.jl

```diff
@@ -384,3 +384,33 @@ macro NamedTuple(ex)
     types = [esc(e isa Symbol ? :Any : e.args[2]) for e in decls]
     return :(NamedTuple{($(vars...),), Tuple{$(types...)}})
 end
+
+"""
+    namedtuple(K, V)
+
+Creates a `NamedTuple` object out of `K`, an iterator of keys and `V`, an iterator of values.  All
+keys must be of type `Symbol`.
+
+# Examples
+```jldoctest
+julia> namedtuple([:a, :b, :c], [1, 2, 3])
+ (a = 1, b = 2, c = 3)
+```
+"""
+namedtuple(K, V) = NamedTuple{Tuple(K)}(V)
+
+"""
+    namedtuple(d)
+
+Create a `NamedTuple` object from any iterable of `Pair`s `d` (for example, any `AbstractDict`).
+The first value in each pair must be of type `Symbol`.
+"""
+namedtuple(d::AbstractDict) = namedtuple(keys(d), values(d))
+function namedtuple(d)
+    NamedTuple{Tuple(first(p) for p ∈ d)}(last(p) for p ∈ d)
```

Comment:
Should probably not iterate `d` twice.

Reply:
agreed, but it seems hard to avoid this without allocating some additional object, since in particular `Tuple` creates an immutable object.  Suggestions?

Reply:
You can just iterate `d` in a for loop and build up two tuples for the names and values by splatting.

Reply:
I strongly suspect that will be slow, but I will try it

Reply:
An alternative might be to call `Tuple` on the whole dict first and then map `first` and `last` over that. That might also make constant propagation easier in some cases.

Reply:
actually wait.. if I do it that way, wouldn't I still have to iterate twice to create the separate tuples?  Can you share an example; sorry if I'm being really thick here... either I'm just being an idiot or this is actually kind of hard to do...

Reply:
Something like:
```julia
keys, vals = (), ()
for (k, v) in d
    keys, vals = (keys..., k), (vals..., v)
end
```
Although I suspect that just `Tuple` and the `map` could be faster.

Reply:
yeah I really can't imagine this being faster, but let me check...

Reply:
It's not about being faster but about being correct for e.g. stateful iterators.

Reply:
yeah, this seems a little bit slower... I'm getting between 800 ns and 900 ns for two elements everything I've tried so far which is undoubtedly incredibly slow.

It seems to me that the problem is that it's just generally hard to create multiple immutable objects in a single loop without creating extra allocations.

Reply:
So far my original version of this scales better with size than anything else I've tried.

Reply:
For me, my second proposal is similarly fast than your original version (on master):
```julia
julia> using BenchmarkTools

julia> function f1(d)
           keys, vals = (), ()
           for (k, v) in d
               keys, vals = (keys..., k), (vals..., v)
           end
           NamedTuple{keys}(vals)
       end
f1 (generic function with 1 method)

julia> function f2(d)
           t = Tuple(d)
           NamedTuple{map(first, t)}(map(last, t))
       end
f2 (generic function with 1 method)

julia> function f3(d)
           NamedTuple{Tuple(first(i) for i in d)}(last(i) for i in d)
       end
f3 (generic function with 1 method)

julia> @btime f1([:a => 1, :b => 2])
  1.407 μs (13 allocations: 464 bytes)
(a = 1, b = 2)

julia> @btime f2([:a => 1, :b => 2])
  1.167 μs (11 allocations: 592 bytes)
(a = 1, b = 2)

julia> @btime f3([:a => 1, :b => 2])
  1.374 μs (10 allocations: 576 bytes)
(a = 1, b = 2)

julia> @btime f1([Symbol(k) => v for (k, v) in zip('a':'z', 1:26)])
  24.818 μs (135 allocations: 8.58 KiB)
(a = 1, b = 2, c = 3, d = 4, e = 5, f = 6, g = 7, h = 8, i = 9, j = 10, k = 11, l = 12, m = 13, n = 14, o = 15, p = 16, q = 17, r = 18, s = 19, t = 20, u = 21, v = 22, w = 23, x = 24, y = 25, z = 26)

julia> @btime f2([Symbol(k) => v for (k, v) in zip('a':'z', 1:26)])
  5.494 μs (63 allocations: 4.41 KiB)
(a = 1, b = 2, c = 3, d = 4, e = 5, f = 6, g = 7, h = 8, i = 9, j = 10, k = 11, l = 12, m = 13, n = 14, o = 15, p = 16, q = 17, r = 18, s = 19, t = 20, u = 21, v = 22, w = 23, x = 24, y = 25, z = 26)

julia> @btime f3([Symbol(k) => v for (k, v) in zip('a':'z', 1:26)])
  4.768 μs (36 allocations: 2.69 KiB)
(a = 1, b = 2, c = 3, d = 4, e = 5, f = 6, g = 7, h = 8, i = 9, j = 10, k = 11, l = 12, m = 13, n = 14, o = 15, p = 16, q = 17, r = 18, s = 19, t = 20, u = 21, v = 22, w = 23, x = 24, y = 25, z = 26)
```

Reply:
(interestingly, 1.5 is quite a bit faster here:
```julia
julia> @btime f1([:a => 1, :b => 2])
  975.231 ns (13 allocations: 464 bytes)
(a = 1, b = 2)

julia> @btime f2([:a => 1, :b => 2])
  691.838 ns (11 allocations: 592 bytes)
(a = 1, b = 2)

julia> @btime f3([:a => 1, :b => 2])
  867.345 ns (10 allocations: 576 bytes)
(a = 1, b = 2)
```
)

Reply:
I wonder if it makes more sense to benchmark bigger dictionaries rather than very small ones? Since it's an inherently type-unstable operation, so you'll be hit by dynamic dispatch for the next operation you do with the result, which means the speed here only matters when it's comparable to the cost of that dispatch

Reply:
I think with much more than 26 entries, you shouldn't really be using a `NamedTuple` though.

Reply:
true, I hadn't seen those at the end of your benchmarks

Reply:
Yeah, sorry, I'm actually on 1.5, I just never bothered building master.

Alright, seems like we have uncovered a significant performance regression.  I'm seeing similar results to what @simeonschaub is showing in that `f1`.

It looks like `f2` is the best option right now, so I'm changing it to that, commit incoming...